osu.Game/Online/Solo/SoloStatisticsWatcher.cs

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using osu.Framework.Allocation;
+using osu.Framework.Extensions.ObjectExtensions;
+using osu.Framework.Graphics;
+using osu.Game.Online.API;
+using osu.Game.Online.API.Requests;
+using osu.Game.Online.API.Requests.Responses;
+using osu.Game.Online.Spectator;
+using osu.Game.Scoring;
+using osu.Game.Users;
+
+namespace osu.Game.Online.Solo
+{
+    /// <summary>
+    /// A persistent component that binds to the spectator server and API in order to deliver updates about the logged in user's gameplay statistics.
+    /// </summary>
+    public partial class SoloStatisticsWatcher : Component
+    {
+        [Resolved]
+        private SpectatorClient spectatorClient { get; set; } = null!;
+
+        [Resolved]
+        private IAPIProvider api { get; set; } = null!;
+
+        private readonly Dictionary<long, StatisticsUpdateCallback> callbacks = new Dictionary<long, StatisticsUpdateCallback>();
+        private long? lastProcessedScoreId;
+
+        private readonly Dictionary<string, UserStatistics> latestStatistics = new Dictionary<string, UserStatistics>();
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            api.LocalUser.BindValueChanged(user => onUserChanged(user.NewValue), true);
+            spectatorClient.OnUserScoreProcessed += userScoreProcessed;
+        }
+
+        /// <summary>
+        /// Registers for a user statistics update after the given <paramref name="score"/> has been processed server-side.
+        /// </summary>
+        /// <param name="score">The score to listen for the statistics update for.</param>
+        /// <param name="onUpdateReady">The callback to be invoked once the statistics update has been prepared.</param>
+        public void RegisterForStatisticsUpdateAfter(ScoreInfo score, Action<SoloStatisticsUpdate> onUpdateReady) => Schedule(() =>
+        {
+            if (!api.IsLoggedIn)
+                return;
+
+            var callback = new StatisticsUpdateCallback(score, onUpdateReady);
+
+            if (lastProcessedScoreId == score.OnlineID)
+            {
+                requestStatisticsUpdate(api.LocalUser.Value.Id, callback);
+                return;
+            }
+
+            callbacks[score.OnlineID] = callback;
```

Comment:
Just for sanity, shall we `Debug.Assert` or `throw` if attempting to register for callback when one already exists? Right now it will silently overwrite.

Reply:
I can do `.Add()` which will have the same effect.